examples/ed25519/src/ed25519.rs

```diff
@@ -0,0 +1,909 @@
+
+    let x_s = (p[31usize] & U8(128u8)) >> 7;
+    let mut y_s = p;
+    y_s[31usize] = y_s[31usize] & U8(127u8);
+    let y = Ed25519FieldElement::from_byte_seq_le(y_s);
+    let z = Ed25519FieldElement::ONE();
+    let yy = y * y;
+    let u = yy - z;
+    let v = d * yy + z;
+    let xx = u * v.inv();
+    let mut x = sqrt(xx)?;
+    let x_r = is_negative(x);
+    if x_r.declassify() != x_s.declassify() {
+        x = Ed25519FieldElement::ZERO() - x;
+    }
+    Some((x, y, z, x * y))
+}
+
+fn point_add(p: EdPoint, q: EdPoint) -> EdPoint {
+    let d_c = Ed25519FieldElement::from_byte_seq_le(CONSTANT_D);
+    let two = Ed25519FieldElement::TWO();
+
+    let (x1, y1, z1, t1) = p;
+    let (x2, y2, z2, t2) = q;
+    let a = (y1 - x1) * (y2 - x2);
+    let b = (y1 + x1) * (y2 + x2);
+    let c = t1 * two * d_c * t2;
+    let d = z1 * two * z2;
+    let e = b - a;
+    let f = d - c;
+    let g = d + c;
+    let h = b + a;
+    let x3 = e * f;
+    let y3 = g * h;
+    let t3 = e * h;
+    let z3 = f * g;
+    (x3, y3, z3, t3)
+}
+
+fn point_identity() -> EdPoint {
+    (
+        Ed25519FieldElement::ZERO(),
+        Ed25519FieldElement::ONE(),
+        Ed25519FieldElement::ONE(),
+        Ed25519FieldElement::ZERO(),
+    )
+}
+
+fn point_mul(s: Scalar, p: EdPoint) -> EdPoint {
+    let mut p = p;
+    let mut q = point_identity();
+    for i in 0..256 {
+        if s.bit(i) {
+            q = point_add(q, p);
+        }
+        p = point_add(p, p);
+    }
+    q
+}
+
+fn point_mul_by_cofactor(p: EdPoint) -> EdPoint {
+    let p2 = point_add(p, p);
+    let p4 = point_add(p2, p2);
+    let p8 = point_add(p4, p4);
+    p8
+}
+
+fn point_neg(p: EdPoint) -> EdPoint {
+    let (x, y, z, t) = p;
+    (
+        Ed25519FieldElement::ZERO() - x,
+        y,
+        z,
+        Ed25519FieldElement::ZERO() - t,
+    )
+}
+
+fn point_eq(p: EdPoint, q: EdPoint) -> bool {
+    let (x1, y1, z1, _) = p;
+    let (x2, y2, z2, _) = q;
+    x1 * z2 == x2 * z1 && y1 * z2 == y2 * z1
+}
+
+fn secret_expand(sk: SecretKey) -> (SerializedScalar, SerializedScalar) {
+    let h = sha512(&ByteSeq::from_slice(&sk, 0, 32));
+    let h_d = SerializedScalar::from_slice(&h, 32, 32);
+    let mut s = SerializedScalar::from_slice(&h, 0, 32);
+    s[0usize] = s[0usize] & U8(248u8);
+    s[31usize] = s[31usize] & U8(127u8);
+    s[31usize] = s[31usize] | U8(64u8);
+    (s, h_d)
+}
+
+pub fn secret_to_public(sk: SecretKey) -> PublicKey {
+    let (s, _) = secret_expand(sk);
+    let base = decompress(BASE).unwrap();
+    let ss = Scalar::from_byte_seq_le(s);
+    let a = point_mul(ss, base);
+    compress(a)
+}
+
+fn check_canonical_scalar(s: SerializedScalar) -> bool {
+    if (s[31usize] & U8(224u8)).declassify() != 0u8 {
+        false
+    } else {
+        BigInteger::from_byte_seq_le(s) < BigInteger::from_byte_seq_le(CONSTANT_L)
+    }
+}
+
+fn scalar_from_hash(h: Sha512Digest) -> Scalar {
+    let s = BigScalar::from_byte_seq_le(h);
```

Comment:
I was wondering where you need `BigScalar`. But here you do indeed. Ideally there should be afunction to convert bytes to natural numbers modulo a value without this detour. Nothing to fix here, but a todo for the library.